### ISSUE TITLE
Typing "#if" unfolds collapsed code

### ISSUE THREAD

**Issue by opopop** [OmniSharp/omnisharp-vscode#3217]
Issue Type: <b>Bug</b>

Typing '#if' immediately unfolds code from cursor to the end of file. Have to fold it back by hand everytime after this :/
I checked some other preprocessor directives, seems like they ok, but typing '/*' causes the issue too.
Screeenshots:
http://prntscr.com/opdjz7
http://prntscr.com/opdk44

Extension version: 1.21.0
VS Code version: Code 1.36.1 (2213894ea0415ee8c85c5eea0d0ff81ecc191529, 2019-07-08T22:59:35.033Z)
OS version: Windows_NT x64 10.0.17763

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-3930K CPU @ 3.20GHz (12 x 3200)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: disabled_software<br>oop_rasterization: disabled_off<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_deferred_display_list: disabled_off<br>skia_renderer: disabled_off<br>surface_synchronization: enabled_on<br>video_decode: enabled<br>viz_display_compositor: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.92GB (12.62GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details>

**Comment by rchande** [comment]
My guess for what happens here is VS Code requests new folding every time you type. If you have typed #if, but no the close directive, our folding provider gets confused and does not return any regions.

**Comment by opopop** [comment]
It seems like that. But "#region" apparently works differently and does not create this problem. It would be nice to have this fixed because it is very inconvenient especially when working with large files.

**Comment by rchande** [comment]
@opopop I took a closer look. This is basically an artifact of error recovery in the compiler. When you type `#if` without `true` or `false`, it looks like the parser implicitly adds `true` and considers the `#endif` to appear at the end of the file. Areas of the file that are #if'd out aren't parsed, and we therefore see no folding spans after #if. Because VS Code updates the folding regions when you type the #if, we return the folding data based on the error recovery, which is now no regions, causing everything to unfold. 

I filed https://github.com/dotnet/roslyn/issues/37910 against Roslyn to track this--feel free to follow up there and add further thoughts/comments.
